### DISCUSSION TITLE
Dependencies are not mapped or passed properly?

### DISCUSSION THREAD

**Discussion by IbrahimS2** [breezefront/community#288] [1. Help] [answered]
I am having a very strange issue, and I cannot make sense of it, I have been working on this integration for a while, but one of the dependencies not getting passed for some reason, any hint is appreciated.

`breeze_default.xml`
```
<item name="Mirasvit_Tm/js/event/event-selectors" xsi:type="array">
    <item name="path" xsi:type="string">Vendor_BreezeMirasvitGTM/js/event/event-selectors</item>
    <item name="import" xsi:type="array">
        <item name="parent" xsi:type="string">Mirasvit_Tm/js/event/event-selectors</item>
    </item>
</item>

<item name="Mirasvit_Tm/js/event/abstract" xsi:type="array">
    <item name="path" xsi:type="string">Vendor_BreezeMirasvitGTM/js/event/abstract</item>
    <item name="import" xsi:type="array">
        <item name="event-selectors" xsi:type="string">Mirasvit_Tm/js/event/event-selectors</item>
        <item name="parent" xsi:type="string">Mirasvit_Tm/js/event/abstract</item>
    </item>
</item>
```

`Vendor_BreezeMirasvitGTM/js/event/abstract`

```
define(["uiComponent"], function (Component) {
    "use strict";

    Component.register("Mirasvit_Tm/js/event/abstract");
});
```

`Mirasvit_Tm/js/event/abstract`
```
define([
    'jquery',
    'uiComponent',
    'underscore',
    'Mirasvit_Tm/js/event/event-selectors'
], function ($, Component, _, EventSelectors) {
    'use strict';

    return Component.extend({
        gtmAttrRegex: /^data-gtm-(.*)/g,

        eventAttr:    'data-gtm-event',
        listAttr:     'data-gtm-item_list_id',
        listNameAttr: 'data-gtm-item_list_name',
        productAttr:  'data-gtm-item_id',

        eventSelector:  '',
        listSelector:  '',
        productSelector:  '',

        sentLists: [],

        initialize: function () {
            this.gtmAttrRegex    = EventSelectors.getAttributeRegex();
            this.eventAttr       = EventSelectors.getEventAttribute();
            this.listAttr        = EventSelectors.getListAttribute();
            this.listNameAttr    = EventSelectors.getListNameAttribute();
            this.productAttr     = EventSelectors.getProductAttribute();
            this.listSelector    = EventSelectors.getListSelector();
            this.productSelector = EventSelectors.getProductSelector();

            this._super();

            this.initEventSelector();

            window.dataLayer = window.dataLayer || [];

            this.listeners();
        },

        initEventSelector: function () {
            this.eventSelector = EventSelectors.getEventSelector();
        },

        listeners: function () {

        },

        sendData: function ($el) {

        },

        setGaData: function ($listEl, productIds) {

        },

        isVisible: function ($el) {
            if (!$el.is(":visible")) {
                return false;
            }

            const $win = $(window);

            const elementTop = $el.offset().top;
            const elementBottom = elementTop + $el.outerHeight();

            const viewportTop = $win.scrollTop();
            const viewportBottom = viewportTop + $win.height();

            return elementBottom > viewportTop && elementTop < viewportBottom;
        },

        getProductAttribute: function () {
            return this.productAttr;
        },

        getProductSelector: function () {
            return this.productSelector;
        }
    });
});
```

With all of this in mind, I am getting an error about the function not being found even though I made sure it is imported and executed prior to either `product-storage.js` or `abstract.js`:
![image](https://github.com/breezefront/community/assets/19901858/bcf6a71d-fe3b-4928-81ab-11cf5beb7e1d)

**Comment by vovayatsyuk** [comment]
There were issues with `Component.register` method.

1. Try to update module-breeze to 2.17.1 (Just tagged).
2. Make sure that all unknown dependencies of `Mirasvit_Tm/js/event/event-selectors` are resolved too.

**Reply by IbrahimS2** [reply]
@vovayatsyuk Same issue
![image](https://github.com/breezefront/community/assets/19901858/0a0b9b65-4e5d-4d05-b952-c509e53b4436)

**Reply by vovayatsyuk** [reply]
Can you please share your github repo with me?

**Reply by IbrahimS2** [reply]
@vovayatsyuk Here you go https://git.aqeltech.com/aqeltech/magento-2/breezefront/mirasvit-gtm-integration

We are working on lots of integration to ensure it works perfectly with Breezefront and as I might have mentioned it will be published on Breezefront integration if that is okay so everyone can benefit from it

**Reply by vovayatsyuk** [reply]
Hmm.. Original code looks like this:

```js
let $productEls = $(EventSelectors().getProductSelector());
```

But you've changed it to:

```js
let $productEls = $(EventSelectors.getProductSelector());
```

Why?

**Reply by IbrahimS2** [reply]
@vovayatsyuk Because it is not a function, rather I passed the component as mapped in the define

**Reply by vovayatsyuk** [reply]
I see that it's a function on your screenshot.

**Reply by vovayatsyuk** [reply]
How it could be not a function if it's a function in the mirasvit source code?

**Reply by IbrahimS2** [reply]
@vovayatsyuk Right, I do not get it myself, it is just an object and I am trying to access the methods inside in this case the **getProductSelector**

**Reply by vovayatsyuk** [reply]
Take a look at your screenshot:

![EventSelectors](https://github.com/breezefront/community/assets/306080/496d2d28-614f-4ed2-9186-d331c38a6a9f)

It's a function.

**Reply by IbrahimS2** [reply]
@vovayatsyuk I cannot follow, I see what you are saying but I cannot follow. anyhow, by adding the parentheses we are invoking the function but then even though the method is not returning 
![image](https://github.com/breezefront/community/assets/19901858/366c3dde-fb3f-4df3-8075-eb998e982ec4)

**Reply by vovayatsyuk** [reply]
Can you also share the Mirasvit module with me? My version of mirasvit has all of the js inside the Mirasvit_GoogleTagManager module (not Mirasvit_Tm)

**Reply by IbrahimS2** [reply]
@vovayatsyuk Sent

**Reply by vovayatsyuk** [reply]
Thanks! Fixed: https://github.com/breezefront/module-breeze/commit/7a579ce8a2a614fe708fc79d74bda0e531ae9e27

There are some errors when bundling is enabled. Will work on them too.

**Reply by vovayatsyuk** [reply]
Fixed error with bundling: https://github.com/breezefront/module-breeze/commit/b3f484abf07dc1abd7969187fcbc05a8b46fbc6b

**Reply by vovayatsyuk** [reply]
Great job on integration!

p.s. You don't need to integrate select-shipping/payment mixins because they are used at checkout only and Breeze doesn't support checkout page.

You can remove these lines and related files:

```xml
<item name="Mirasvit_GoogleTagManager/js/event/select-shipping" xsi:type="array">
    <item name="path" xsi:type="string">AQELTech_BreezeMirasvitGTM/js/event/select-shipping</item>
    <item name="import" xsi:type="array">
        <item name="abstract" xsi:type="string">Mirasvit_Tm/js/event/abstract</item>
        <item name="parent" xsi:type="string">Mirasvit_GoogleTagManager/js/event/select-shipping</item>
    </item>
</item>

<item name="AQELTech_BreezeMirasvitGTM/js/event/select-shipping-mixin" xsi:type="array">
    <item name="path" xsi:type="string">AQELTech_BreezeMirasvitGTM/js/event/select-shipping-mixin</item>
    <item name="import" xsi:type="array">
        <item name="select-shipping" xsi:type="string">Mirasvit_GoogleTagManager/js/event/select-shipping</item>
    </item>
</item>

<item name="Mirasvit_Tm/js/event/select-payment" xsi:type="array">
    <item name="path" xsi:type="string">AQELTech_BreezeMirasvitGTM/js/event/select-payment</item>
    <item name="import" xsi:type="array">
        <item name="abstract" xsi:type="string">Mirasvit_Tm/js/event/abstract</item>
        <item name="parent" xsi:type="string">Mirasvit_Tm/js/event/select-payment</item>
    </item>
</item>

<item name="AQELTech_BreezeMirasvitGTM/js/event/select-payment-mixin" xsi:type="array">
    <item name="path" xsi:type="string">AQELTech_BreezeMirasvitGTM/js/event/select-payment-mixin</item>
    <item name="import" xsi:type="array">
        <item name="select-payment" xsi:type="string">Mirasvit_Tm/js/event/select-payment</item>
    </item>
</item>
```

**Reply by IbrahimS2** [reply]
@vovayatsyuk Can we publish this integration so others can make use of it? I will publish it to the packager to install it via composer.

And about my work with the EventsSelector, do you have any comments?

**Reply by IbrahimS2** [reply]
@vovayatsyuk 
![image](https://github.com/breezefront/community/assets/19901858/cf467810-3019-491f-a664-3ba1101b2711)
![image](https://github.com/breezefront/community/assets/19901858/3a169230-2b2b-4dfd-85ef-7896c057af4b)

**Reply by vovayatsyuk** [reply, marked as answer]
I fixed this error in https://github.com/breezefront/module-breeze/commit/7a579ce8a2a614fe708fc79d74bda0e531ae9e27

**Reply by IbrahimS2** [reply]
@vovayatsyuk I already upgraded to the latest version

**Reply by vovayatsyuk** [reply]
Please note, that these two commits are not tagged yet. Ensure that you have these changes in the code. If so and the bug is still here, please share the URL with me.

**Reply by IbrahimS2** [reply]
@vovayatsyuk It is working which is great, thank you. can we publish our integration with GTM on breezefront?

**Reply by vovayatsyuk** [reply]
Yes, sure! You can add your repo to packagist and I will update https://breezefront.com/extensions page.

**Reply by IbrahimS2** [reply]
@vovayatsyuk `aqeltech/module-breeze-mirasvit-gtm`

**Reply by vovayatsyuk** [reply]
Done!

p.s. These patches are not needed too: https://git.aqeltech.com/aqeltech/magento-2/breezefront/mirasvit-gtm-integration#required-patches.
